addons/fleet/models/fleet_vehicle_odometer.py

```diff
@@ -30,3 +31,42 @@ def _compute_vehicle_log_name(self):
     def _onchange_vehicle(self):
         if self.vehicle_id:
             self.unit = self.vehicle_id.odometer_unit
+
+    def _check_odometer(self):
+        vehicle_ids = list(set(self.vehicle_id.ids))
+        odometers_to_check = self.search([('vehicle_id', 'in', vehicle_ids)], order='date')
+
+        grouped_odometers = {}
+        for odometer in odometers_to_check:
+            if not grouped_odometers.get(odometer.vehicle_id.id):
+                grouped_odometers[odometer.vehicle_id.id] = odometer
+            else:
+                grouped_odometers[odometer.vehicle_id.id] |= odometer
+
+        for odometer in self:
+            this_vehicle_odometers = grouped_odometers.get(odometer.vehicle_id.id, self.env['fleet.vehicle.odometer'])
+            for odometer_to_check in this_vehicle_odometers:
+                if odometer.date < odometer_to_check.date:
+                    if odometer.value > odometer_to_check.value:
+                        raise UserError(_('You cannot set an odometer value higher than one from the future.'))
+                elif odometer.date == odometer_to_check.date:
+                    if odometer.id > odometer_to_check.id:
+                        if odometer.value < odometer_to_check.value:
+                            raise UserError(_('You cannot set an odometer value lower than one from a previous record.'))
+                    elif odometer.id < odometer_to_check.id:
+                        if odometer.value > odometer_to_check.value:
+                            raise UserError(_('You cannot set an odometer value higher than one from a future record.'))
+                else:
+                    if odometer.value < odometer_to_check.value:
```

Comment:
Actually, it seems very complicated. 

As you call your check method after super, why don't you just check that:
```py
    def _check_odometer(self):
        def is_ascending(odometers):
            if not odometers:
                return True
            return all(a <= b for a, b in zip(odometers, odometers[1:]))

        odometers_to_check = self.vehicle_id.odometer_ids.sorted("date")

        odometers_by_vehicle = defaultdict(lambda: self.env['fleet.vehicle.odometer'])
        for odometer in odometers_to_check:
            odometers_by_vehicle[odometer.vehicle_id] += odometer

        for vehicle, odometers in odometers_by_vehicle.items():
            if not is_ascending(odometers.mapped('value')):
                raise UserError(_('The odometers values for vehicle %s are not ascending', vehicle.display_name))
```